### ISSUE TITLE
Working with time-series data?

### ISSUE THREAD

**Issue by sliem** [InvestmentSystems/static-frame#30]
Hi, got my interest peaked by your interview with Micheal Kennedy at Talk Python to Me. The library looks very interesting, especially the focus on removing ambiguity. 

Most of my pandas work involve handling time-series data. Specifically of sensor data where the time between isn’t at all constant. I do lots of grouping by various time intervals, and time-aware interpolations.

After an admittedly cursory glance I didn’t see how I would do that kind of work with static-frame. Am I missing something, or haven’t you had reason/opportunity to consider that kind of use case?

**Comment by flexatone** [comment]
Thanks for your interest and question.

Right now, we have Index sub classes for working with time series at the frequency days, months, and years. I assume what you would need would be at the level of time stamps. 

StaticFrame uses NumPy date time types to implement these indices, and there is support for timestamps, we just have not yet implemented anything at the sub-day level yet.

https://docs.scipy.org/doc/numpy-1.15.0/reference/arrays.datetime.html

I would be interested in seeing a small example of how you work with time-series data: i.e., what format you are reading in, what sort of resolution you need, etc. We should be able to at least support a timestamp level index easily. We do not yet have any interpolation routines built-in, but seeing some examples of what you do might give us some ideas.

**Comment by alkasm** [comment]
I too work with irregularly sampled time-series sensor data. 

Generally I feed the results of a SQL query directly into Pandas; the timestamps are stored in ISO format strings, and then I use `pd.to_datetime()` to convert them into `datetime64[ns]` (I don't actually use nanosecond resolution, but that's what Pandas gives me by default).

I also use `pd.to_timedelta()` as well, which uses the `np.timedelta64` type. 

My loading regime will often look something like this:

        df = pd.read_csv(csvfile) 
        for col in datetime_cols:
            df[col] = pd.to_datetime(df[col])

        for col in timedelta_cols:
            df[col] = pd.to_timedelta(df[col], unit="ms")

        df = df.set_index(["timestamp", "row_index"])

**Comment by flexatone** [comment]
Many thanks for this example; I will explore how to do something similar with StaticFrame.

**Comment by flexatone** [comment]
I have implemented two new Index subclasses (to be released in 0.3.3) that provide at least basic support with working with intra-day datetime indices: `IndexSecond` and `IndexMillisecond`. Both of these are backed by np.datetime64 types, and the choice between the two determines (a) minimum resolution interpreted and (b) relevance of integers used in construction (for `IndexSecond`, it is seconds from the epoch; for `IndexMillisecond`, it is milliseconds from the epoch).

For example, using the T0.csv file from here:
https://www.kaggle.com/BGU-CSRC/sherlock

We can load in a Frame (via pandas `read_csv()`), set a column as an IndexMillisecond, and then use that index to do partial selection based on strings.

I will close this issue; please open new issues for additional specific features based on these foundations. 

```python
>>> f1 = sf.Frame.from_pandas(pd.read_csv('T0.csv'))
>>> f2 = f1.set_index('SystemInfo_timestemp', index_constructor=sf.IndexMillisecond, drop=True)[f1.columns.iloc[:4]]

>>> f2.loc['2016-04']
<Frame>
<Index>                 UserId     UUID          Version  Telephoney_deviceId  <object>
<IndexMillisecond>
2016-04-28T04:22:12.226 97bb95f55a 1461810132306 2.2      0xC9BBAB88E558ED7...
2016-04-28T16:29:21.320 97bb95f55a 1461853761332 2.2      0xC9BBAB88E558ED7...
2016-04-28T17:36:13.733 97bb95f55a 1461854175049 2.2      0xC9BBAB88E558ED7...
2016-04-30T20:21:07.848 97bb95f55a 1462040467954 2.2      0xC9BBAB88E558ED7...
<datetime64[ms]>        <object>   <int64>       <object> <object>

>>> f2.loc['2016-05-16T00:29']
<Frame>
<Index>                 UserId     UUID          Version  Telephoney_deviceId  <object>
<IndexMillisecond>
2016-05-16T00:29:50.559 97bb95f55a 1463347793901 2.3.2    0xC9BBAB88E558ED7...
2016-05-16T00:29:54.465 97bb95f55a 1463351394569 2.3.2    0xC9BBAB88E558ED7...
<datetime64[ms]>        <object>   <int64>       <object> <object>
```
